### DISCUSSION TITLE
Visibility of `JVisaInstrument.readBytes(int bufferSize)` method

### DISCUSSION THREAD

**Discussion by penguinlab** [pfroud/JVisa#10] [General]
I was wondering why the visibility of the `readBytes(int bufferSize)` method on `JVisaInstrument` is protected. Is there any reason for this?

Other methods for reading responses (`readString()`, `readString(int bufferSize)`) are public.
Therefore, I feel that we may as well make it public in the same way.

**Comment by pfroud** [comment]
Agreed. Do you use the `readBytes()` method for anything? I think it is only useful for debugging, and usually it would be more useful to have a string be returned.

I also want to make these changes for JVisa version 2:

* Remove `sendAndReceiveBytes()` because I think it is useless
* Change `sendAndReceiveString()` to `query()`

Do you ever use `sendAndReceiveBytes()`?

**Reply by penguinlab** [reply]
Thanks for changing this to discussion, it wasn't appropriate to put it in the issue.

I did not use readBytes(), but since only this method was protected, I thought there might be a reason.

Some instruments have internal storage and file systems, and commands are provided to retrieve files.
Therefore, I think you should keep sendAndReceiveBytes().

In the software I am developing, I am automating taking a screenshot of a Keysight N9323C Spectrum Analyzer and getting its file, and that is where I use sendAndReceiveBytes().

**Comment by pfroud** [comment]
Oh, I didn't know about file systems and screenshots! I will keep the `sendAndReceiveBytes()` method (I think I will rename it `queryBytes()`)

**Reply by penguinlab** [reply]
I think it is a good and concise method name.
It looks good alongside other existing methods.

**Reply by pfroud** [reply]
Done in commit 46511271fcd646d4d4560e3de461317665c79114

**Comment by penguinlab** [comment]
One additional suggestion: the default response buffer size for `sendAndReceiveBytes()` is 1024 bytes, but it could be made somewhat larger.

In the case of string responses, I don't think there are any cases where the size exceeds 1024 bytes, but in the case of binary responses (i.e., when retrieving a file as mentioned in the previous comment), the response will be larger.

However, since `sendAndReceiveBytes(int bufSize)` is also available to specify the buffer size as an argument, I'm not troubled at present.

**Comment by pfroud** [comment]
Looks like PyVISA uses a buffer size ([`chunk_size`](https://github.com/pyvisa/pyvisa/blob/1.13.0/pyvisa/resources/messagebased.py#L650)) of 20 * 1024 bytes which is a lot.

Also, PyVISA has a function called [`read_bytes()`](https://github.com/pyvisa/pyvisa/blob/1.13.0/pyvisa/resources/messagebased.py#L325) where you can specify the total size and the chunk size:

```
Parameters
----------
count : int
    The number of bytes to read from the instrument.
chunk_size : Optional[int], optional
    The chunk size to use to perform the reading. If count > chunk_size
    multiple low level operations will be performed. Defaults to None,
    meaning the resource wide set value is set.
```

It would be interesting to add that to JVisa. I have access to an oscilloscope which returns binary data from a screenshot request so I can try it.

**Reply by penguinlab** [reply]
When retrieving file data for the Keysight N9323C, the response byte sequence contained 8 bytes of unnecessary data at the beginning.

This is probably an instrument-specific specification. Please note that there may be other instruments that return a similar response.

**Reply by pfroud** [reply]
Was the first byte 0x23 (decimal 35) (ASCII `#` character)? If yes, it is probably a header which can tell you the length of the rest of the data

https://www.mathworks.com/matlabcentral/answers/1598789-what-is-the-data-format-used-by-readbinblock-and-writebinblock-functions-in-instrument-control-toolb#answer_843289

**Reply by penguinlab** [reply]
Oh, I see! The first byte was 0x23.
Thanks for the good info.

**Comment by pfroud** [comment]
Here is more information about the header. It is defined in IEEE 488.2-1992. I found a copy of the online [here](http://webuser.unicas.it/misure/MAQ_OLD%20(VO)/Dispense/DISP_7STANDARD%20IEEE%20488_2%201992.pdf) but that link might not work in the future.

There are two relevant sections:

* 7.7.6 &lt;ARBITRARY BLOCK PROGRAM DATA&gt; (PDF page 84)
* 8.7.9 &lt;DEFINITE LENGTH ARBITRARY BLOCK RESPONSE DATA&gt; (PDF page 102)

Diagram from section 8.7.9.2:

![definite length arbitrary data block response data](https://user-images.githubusercontent.com/9056906/220241741-27a515dd-65da-4c93-b026-2b75089be305.png)

Description from section 7.7.6.4:

> The value of the &lt;nonzero digit&gt; element shall equal the number of &lt;digit&gt; elements that follow. The value of the &lt;digit&gt; elements taken together as a decimal integer shall equal the number of &lt;8 bit data byte&gt; elements that follow.

&nbsp;

I learned about this a few days ago. I was reading the [programming manual](https://beyondmeasure.rigoltech.com/acton/attachment/1579/f-0386/1/-/-/-/-/DS1000Z_Programming%20Guide_EN.pdf) for the Rigol DS1000Z series oscilloscope. To take a screenshot, you send `display:data?` and the instrument returns image data with a header at the beginning (PDF page 79).

The programming manual calls the header "TMC Blockheader." I think TMC means the USB Test & Measurement Class, but I couldn't find other references to TMC Blockheader.

I wanted to learn if the header format was made by Rigol or if it was a standard format. I researched what format is used by other instruments to return screenshots. I found a different manufacturer call it "488 block data" then found the source.

**Reply by penguinlab** [reply]
Thanks for the detailed info.
I looked at the documentation you gave me, and it looks like the header size is not always 8 bytes.

It seems to me that `sendAndReceiveBytes()` (and `queryBytes()`) should not process the headers, but simply return the byte sequence of the response untouched.

**Reply by pfroud** [reply]
Agreed, the header can be variable length, so the `queryBytes()` method should not do anything. Maybe I will add a method called `readBinaryBlock()` or something

**Reply by pfroud** [reply]
Support for reading binary blocks added in release [2.0.0](https://github.com/pfroud/JVisa/releases/tag/2.0.0)
